### PR TITLE
[SPARK-21619][SQL] Fail the execution of canonicalized plans explicitly

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?
Canonicalized plans are not supposed to be executed. I ran into a case in which there's some code that accidentally calls execute on a canonicalized plan. This patch throws a more explicit exception when that happens.

## How was this patch tested?
Added a test case in SparkPlanSuite.

### CODE REVIEW

**Comment by rxin** [comment]
cc @adrian-ionescu @gatorsmile

**Comment by gatorsmile** [comment]
test this please

**Comment by gatorsmile** [comment]
cc @shaneknapp It sounds like we are unable to trigger the test. Could you please check it?

**Review by gatorsmile — APPROVED** [review]
LGTM pending Jenkins.

**Comment by rxin** [comment]
Still looking into it, but the failure is related to reuse exchange and caching.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 1]
```diff
@@ -181,17 +181,38 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
   override def innerChildren: Seq[QueryPlan[_]] = subqueries
 
   /**
+   * A private mutable variable to indicate whether this plan is the result of canonicalization.
+   * This is used solely for making sure we wouldn't execute a canonicalized plan.
+   * See [[canonicalized]] on how this is set.
+   */
+  @transient
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: I guess plans are already not valid on executors, by why `@transient`?

**Comment by gatorsmile** [comment]
There is another zero argument  [ResetCommand](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala#L155)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SparkPlanSuite.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  test("SPARK-21619 execution of a canonicalized plan should fail") {
+    val plan = spark.range(10).queryExecution.executedPlan.canonicalized
+
+    intercept[IllegalStateException] { plan.execute() }
+    intercept[IllegalStateException] { plan.executeCollect() }
+    intercept[IllegalStateException] { plan.executeCollectPublic() }
+    intercept[IllegalStateException] { plan.executeToIterator() }
+    intercept[IllegalStateException] { plan.executeBroadcast() }
+    intercept[IllegalStateException] { plan.executeTake(1) }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SparkPlanSuite.scala]: nit. There is an inconsistent corner case in `plan.executeTake`.
```scala
plan.executeTake(1)  -> raise exception
plan.executeTake(0)  -> no exception
plan.executeTake(-1)  -> raise exception
```

**Comment by gatorsmile** [comment]
ping @rxin

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SparkPlanSuite.scala (hunk 2 above)
Comment (rxin) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SparkPlanSuite.scala]: That's not an issue with this test, is it? It's just how execution is done.

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**